src/Extension/AttributeExtension.php

```diff
@@ -0,0 +1,179 @@
+                        'needs_context' => $needsContext,
+                        'is_variadic' => $isVariadic,
+                        'is_safe' => $attribute->isSafe,
+                        'is_safe_callback' => $attribute->isSafeCallback,
+                        'pre_escape' => $attribute->preEscape,
+                        'preserves_safety' => $attribute->preservesSafety,
+                        'deprecated' => $attribute->deprecated,
+                        'alternative' => $attribute->alternative,
+                    ]);
+                }
+
+                // Functions
+                foreach ($method->getAttributes(AsTwigFunction::class) as $attribute) {
+                    $attribute = $attribute->newInstance();
+
+                    $name = $attribute->name;
+                    if (isset($functions[$name])) {
+                        throw new \LogicException(sprintf('Multiple definitions of the "%s" function.', $name));
+                    }
+
+                    $parameters = $method->getParameters();
+                    $needsEnvironment = isset($parameters[0]) && Environment::class === $parameters[0]->getType()?->getName();
+                    $firstParam = $needsEnvironment ? 1 : 0;
+                    $needsContext = isset($parameters[$firstParam]) && 'context' === $parameters[$firstParam]->getName() && 'array' === $parameters[$firstParam]->getType()?->getName();
+                    $firstParam += $needsContext ? 1 : 0;
+                    $isVariadic = isset($parameters[$firstParam]) && end($parameters)->isVariadic();
+
+                    $functions[$name] = new TwigFunction($name, [$objectOrClass, $method->getName()], [
+                        'needs_environment' => $needsEnvironment,
+                        'needs_context' => $needsContext,
+                        'is_variadic' => $isVariadic,
+                        'is_safe' => $attribute->isSafe,
+                        'is_safe_callback' => $attribute->isSafeCallback,
+                        'deprecated' => $attribute->deprecated,
+                        'alternative' => $attribute->alternative,
+                    ]);
+                }
+
+                // Tests
+                foreach ($method->getAttributes(AsTwigTest::class) as $attribute) {
+                    $attribute = $attribute->newInstance();
+
+                    $name = $attribute->name;
+                    if (isset($tests[$name])) {
+                        throw new \LogicException(sprintf('Multiple definitions of the "%s" test.', $name));
+                    }
+
+                    $parameters = $method->getParameters();
+                    $isVariadic = isset($parameters[$firstParam]) && end($parameters)->isVariadic();
+
+                    $tests[$name] = new TwigTest($name, [$objectOrClass, $method->getName()], [
+                        'is_variadic' => $isVariadic,
+                        'deprecated' => $attribute->deprecated,
```

Comment:
\+ options `node_class` and `one_mandatory_argument` ?

 (if TwigTest keeps it [current signature](https://github.com/twigphp/Twig/blob/aeeec9a5e907a79e50a6bb78979154599401726e/src/TwigTest.php#L38))